features/distributed/coordination/zookeeper/src/test/java/org/opennms/features/distributed/coordination/zookeeper/TestZookeeperDomainManager.java

```diff
@@ -0,0 +1,148 @@
+    private final AtomicInteger connectionAttempts = new AtomicInteger(0);
+    private final AtomicInteger disconnectionAttempts = new AtomicInteger(0);
+    private final List<CompletableFuture<Void>> activeFutures = new CopyOnWriteArrayList<>();
+    private final List<CompletableFuture<Void>> standbyFutures = new CopyOnWriteArrayList<>();
+    private ZookeeperDomainManager manager;
+
+    /**
+     * Test that the registered callbacks are called when ZooKeeper informs the manager to take leadership. Also tests
+     * that only one connection attempt to ZooKeeper was initiated even though multiple registrations/deregistrations
+     * are occurring in parallel.
+     *
+     * @throws InterruptedException
+     */
+    @Test
+    public void testCallbacks() throws InterruptedException, TimeoutException, ExecutionException {
+        CuratorFramework mockClient = mock(CuratorFramework.class);
+        CuratorFrameworkFactory.Builder mockBuilder = mock(CuratorFrameworkFactory.Builder.class);
+        LeaderSelector mockLeaderSelector = mock(LeaderSelector.class);
+
+        when(mockBuilder.build()).thenReturn(mockClient);
+        doAnswer((invocation) -> connectionAttempts.incrementAndGet()).when(mockClient).start();
+        doAnswer((invocation) -> disconnectionAttempts.incrementAndGet()).when(mockClient).close();
+
+        manager = ZookeeperDomainManager.withMocks("test.domain", mockBuilder, mockLeaderSelector);
+        LeaderSelectorListener leaderSelectorListener = manager.getLeaderSelectorListener();
+
+        int numToRegister = 100;
+        IntStream.range(0, numToRegister).parallel().forEach(this::register);
+
+        // We want to wait until the connection from above actually occurred
+        await().atMost(10, TimeUnit.SECONDS).until(manager::isConnected);
+        // Wait some extra time to avoid a race condition where a thread was connecting right after we became connected
+        Thread.sleep(1000);
+        assertEquals(1, connectionAttempts.get());
+
+        // Simulate ZooKeeper telling our client it is leader
+        CompletableFuture<Void> leadershipFuture = CompletableFuture.runAsync(() -> {
+            try {
+                leaderSelectorListener.takeLeadership(null);
+            } catch (Exception e) {
+                throw new CompletionException(e);
+            }
+        });
+
+        // Every registrant should have went active
+        CompletableFuture.allOf(activeFutures.toArray(new CompletableFuture[numToRegister])).get(10,
+                TimeUnit.SECONDS);
+
+        if (leadershipFuture.isDone() || leadershipFuture.isCompletedExceptionally()) {
+            fail("Leadership thread did not block");
+        }
+
+        try {
+            leaderSelectorListener.stateChanged(null, ConnectionState.LOST);
+            fail("No cancel leadership exception was caught");
+        } catch (Exception e) {
+            assertTrue(e instanceof CancelLeadershipException);
```

Comment:
Can we please not use `assertTrue`?
I find it very hard to investigate failing `assertTrue/False` tests as they not clearly state what the problem is. 
I'd opt for either using `assertThat(...)`, `assertEquals(...)` or if you absolutely must use `assertTrue/False`, add the expected condition: `assertTrue("Exception is not a CancelLeadershipException but was expected", e instanceof CancelLeadershipException)`